common: inf and NaN are invalid in JSON

sajibreadd opened the pull request:
Fixes: https://tracker.ceph.com/issues/66215





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

sebastian-philipp commented on src/common/Formatter.cc:
```diff
@@ -296,6 +296,17 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
+void JSONFormatter::add_value(std::string_view name, double val) {
```
wat to add a test to https://github.com/ceph/ceph/blob/main/src/test/common/test_json_formatter.cc ?

sajibreadd replied:
Added the test.

sebastian-philipp commented:
@batrick do you want to review this?

batrick commented on src/test/common/test_json_formatter.cc:
```diff
@@ -79,3 +79,20 @@ TEST(formatter, utime)
   EXPECT_EQ(input.nsec(), output.nsec());
 }
+
+TEST(formatter, bug_66215)
```
```suggestion
TEST(formatter, dump_inf_or_nan)
```

batrick commented on src/common/Formatter.cc:
```diff
@@ -296,6 +296,17 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
+void JSONFormatter::add_value(std::string_view name, double val) {
```
```suggestion
template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
void JSONFormatter::add_value(std::string_view name, T val) {
```

batrick commented on src/common/Formatter.h:
```diff
@@ -232,7 +232,7 @@ namespace ceph {
     void print_name(std::string_view name);
     void print_comma(json_formatter_stack_entry_d& entry);
-
+    void add_value(std::string_view name, double val);
```
```suggestion
template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
    void add_value(std::string_view name, T val);
```

sajibreadd replied:
Shouldn't be it ambiguous if keeping both functions?
```
    template <typename T, typename = std::enable_if_t<std::is_floating_point_v<T>>>
    void add_value(std::string_view name, T val);

    template <typename T>
    void add_value(std::string_view name, T val);
```

batrick replied:
Ya, nevermind.

batrick commented on src/common/Formatter.cc:
```diff
@@ -296,6 +296,17 @@ void JSONFormatter::finish_pending_string()
+void JSONFormatter::add_value(std::string_view name, double val) {
+  CachedStackStringStream css;
+  if (std::isinf(val) || std::isnan(val)) {
+    *css << "null";
```
I would prefer we use a string like "[+-]inf" and "NaN". Then it can be decoded if desired to the original representation.

sajibreadd replied:
Are you looking for quoted string "[+-]inf"?

sebastian-philipp replied:
the JSON RFC doesn't know things like `-inf` or similar. We end up with client libraries that just bail out if they encounter invalid JSON.

sajibreadd replied:
What about quoted string?

patrakov replied:
I would vote for placing such exceptional values into a differently named field, which is, however, impossible to do in this function. In any case, making a field sometimes (especially in rare cases) have a different data type (a string instead of a number) is bad, as it would cause crashes in type-aware clients that implicitly apply a schema.

In short, this function is not a proper place to make a policy on encoding non-numbers in places where numbers are expected.

batrick replied:
Is `null` any better?

patrakov replied:
On paper, yes: with Rust and serde_json, one can directly serialize a "floating point or `null`" into `Option<f32>`.

In another possible opinion, any NaNs or infinities here should be treated as errors (i.e., `ceph_assert`) and all callers fixed up. Of course, this is a monumental task, that's why I am not suggesting this for real, but writing `null` and logging a backtrace might be something to consider.

batrick replied:
Very well, let's stick with `null`.

batrick reviewed (changes requested):
Commit title `JSON Parser does not allow inf and NaN. Converted it to null.` to `common/Formatter: dump inf/nan as strings`

sajibreadd commented:
Is this PR going for the test? Also I think we need to backport this PR.

batrick commented on src/test/common/test_json_formatter.cc:
```diff
@@ -79,3 +79,22 @@ TEST(formatter, utime)
+  formatter.flush(bl);
+  JSONParser parser;
+  cout << std::string(bl.c_str(), bl.length()) << endl;
+  EXPECT_TRUE(parser.parse(bl.c_str(), bl.length()));
```
should we not also verify that the values are `null`?

batrick commented:
jenkins test make check arm64

patrakov commented on src/common/Formatter.cc:
```diff
@@ -296,7 +296,18 @@ void JSONFormatter::finish_pending_string()
-template <class T>
+void JSONFormatter::add_value(std::string_view name, double val) {
+  CachedStackStringStream css;
+  if (std::isinf(val) || std::isnan(val)) {
```
Would the shorter `!std::isfinite(val)` form work here?

sajibreadd replied:
Sorry, missed your comment somehow.
It does work but do you have any special observation for using that instead of `std::isinf(val)` ?

patrakov replied:
It's shorter, expresses the intent more directly, enumerates goodness not badness,... i.e. purely aesthetical reasons.

sajibreadd commented:
> jenkins test make check arm64
@batrick sorry confused a bit. This field is not required. Do I need to do any adjustment to get check mark here?

sajibreadd commented:
jenkins test make check

sajibreadd commented:
jenkins test docs

sajibreadd commented:
jenkins test api
jenkins test windows

sajibreadd commented:
jenkins test api

sajibreadd commented:
jenkins test windows

sajibreadd commented:
jenkins test make check arm64

sajibreadd commented:
jenkins test api

batrick commented on src/common/Formatter.h:
```diff
@@ -232,8 +232,9 @@ namespace ceph {
+    void add_value(std::string_view name, double val);
 
-    template <class T>
+    template <typename T>
```
undo please

batrick commented on src/common/Formatter.cc:
```diff
@@ -296,7 +296,18 @@ void JSONFormatter::finish_pending_string()
   }
 }
 
-template <class T>
```
undo this change

batrick commented on src/test/common/test_json_formatter.cc:
```diff
@@ -79,3 +79,27 @@ TEST(formatter, utime)
+  EXPECT_EQ(parser.find_obj("negative_infinity")->get_data(), "null");
+  EXPECT_EQ(parser.find_obj("nan_val")->get_data(), "null");
+  EXPECT_EQ(parser.find_obj("nan_val_alt")->get_data(), "null");
+}
```
missing terminal newline

ifed01 replied:
please verify double_max/double_min values as well

batrick reviewed (changes requested):
minor nits, otherwise lgtm

sajibreadd commented:
jenkins test make check

sajibreadd commented:
jenkins test make check

sajibreadd commented:
jenkins test api

sajibreadd commented:
jenkins test api

sajibreadd commented:
@batrick Is there anything I need to do in this PR?

batrick commented:
@sajibreadd did you mean to close this? Please reopen?

sajibreadd commented:
> @sajibreadd did you mean to close this? Please reopen?

Here is some mistake I did. I merged main into the branch by mistake(by syncing the fork by mistake). As there is a lot of commit to drop that's why here is the things I have done,
1. created a new branch.
2. cherrypicked the commit.
3. pushed the branch upstream
4. deleted the old branch named wip-66215
5. renamed the newly created branch wip-66215

Now github is not allowing me to reopen the PR. Apologies for the mistakes. Now what should be the best things to do? create another PR and have reference of this PR?
